lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -1320,6 +1320,32 @@ bool LLVMToSPIRV::transDecoration(Value *V, SPIRVValue *BV) {
     }
   }
 
+  if (auto BVF = dyn_cast_or_null<FPMathOperator>(V)) {
+    auto Opcode = BVF->getOpcode();
+    if (Opcode == Instruction::FAdd ||
+        Opcode == Instruction::FSub ||
+        Opcode == Instruction::FMul ||
+        Opcode == Instruction::FDiv ||
+        Opcode == Instruction::FRem) {
```

Comment:
According to https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fp_fast_math_mode_a_fp_fast_math_mode, looks like `OpFMod` is missing, isn't it?

Reply:
There is no `FMod` instruction in LLVM IR.

SPIR-V's `OpFRem` is equivalent to LLVM IR's `FRem` operation, which is equivalent to the C99 remainder operator (`%`). (C90 had a modulus operator, instead, which give an implementation-defined result if one or both of the operands is negative. C99 replaced this with a remainder operator which gives fully language-defined results for negative operands.)

It seems that certain graphics language specs define an operation that is equivalent to `OpFMod`, and it has also been used by some non-graphics languages (e.g. Ada), but my understanding is that we are unlikely to see heavy use of this for OpenCL in the foreseeable future, although support for translation _from_ `OpFMod` is mandated by the SPIR-V specification.

In any case, it translates to a complex sequence of instructions in LLVM IR, of which it may be possible to eliminate some when fast math flags are applied. If we did want to apply fast math flags to instructions generated when translating `OpFMod`, we would want to be selective about which instructions we apply those flags to. Given that `OpFMod` is basically unused for OpenCL, my view is that it's worth being compliant with the specification, but not worth optimizing at present.

Reply:
Note that GLSL has a `mod()` function that is described in GLSL 4.60.7 § 12.2.14. _Mapping of other instructions_ as mapping to `OpFMod`.

The `%` operator is described in that same section as mapping to `OpSMod`, but the description of the `%` operator in § 5.9 _Expressions_ states that "results are undefined if one or both operands are negative", so by my reading, `OpSRem` would be valid, too.

I suppose if we wanted the translator to support GLSL shaders that have been translated into SPIR-V which uses `OpFMod` and/or `OpSMod`, and we wanted to translate those `OpFMod` and/or `OpSMod` instructions into something _efficient_, we would be forced to introduce intrinsics (or _perhaps_ an extra instruction or two in LLVM IR).